### ISSUE TITLE
Accessibility: Self assessment of Status Quo, November

### ISSUE THREAD

**Issue by marcus-herrmann** [owncloud/web#4300]
(This is a "living issue", meaning it will be frequently updated. Also my plan is to use this GitHub issue as an overview, will list every finding that relates to a WCAG/BITV success criterion here. Remediation information on these findings will be in separate GitHub issues)

Date: 11.11.2020 – 18.11.2020

# Methodology
https://testen.bitv-test.de/selbstbewertung/test.php, since it is the closest to the final BITV test

# 1 Scope
Every audit/self assessment needs a scope of pages or app states the audit is limited to. For this self assessment, I'm planning to inspect the following situations regarding accessibility, since I'm considering them the core functionality of OCIS.

* 1.1 Login Page, URL: https://ocis-a11y.owncloud.works/
* 1.2 General Overview when logged in, URL: https://ocis-a11y.owncloud.works/#/files/list
* 1.3 Userflow: Upload a single file, URL: _Process starting at https://ocis-a11y.owncloud.works/#/files/list_
* 1.4 Userflow: Download a single file, URL: _Process starting at https://ocis-a11y.owncloud.works/#/files/list_
* 1.5 Userflow: Share a single file, URL: _Process starting at https://ocis-a11y.owncloud.works/#/files/list_

## Login Page (Scope 1.1)
| BITV test step     | Result                                                                     | Issue | Prio Advice |
| ------------- |------------------------------------------------|--|-|
| 1.1.1a              | not applicable                                                      |||
| 1.1.1b              | not applicable                                                      |||
| 1.1.1c              | not applicable                                                      |||
| 1.1.1d              | not applicable                                                      |||
| 1.2.1a              | not applicable                                                      |||
| 1.2.2a              | not applicable                                                      |||
| 1.2.3a              | not applicable                                                      |||
| 1.2.4a              | not applicable                                                      |||
| 1.2.5a              | not applicable                                                      |||
| 1.3.1a              | not applicable                                                      |||
| 1.3.1b              | not applicable                                                      |||
| 1.3.1c              | not applicable                                                      |||
| 1.3.1d              | not applicable                                                      |||
| 1.3.1e              | not applicable                                                      |||
| 1.3.1f              | not applicable                                                      |||
| 1.3.1g              | not applicable                                                      |||
| 1.3.1h              | Fail 🔴 Inputs have no label, placeholder is not enough | **See #4319**  |++|
| 1.3.2a              | Pass ✅                                                               |||
| 1.3.3a              | Pass ✅                                                               |||
| 1.3.4a              | Pass ✅                                                               |||
| 1.3.5a              | Pass ✅                                                               |||
| 1.4.1a              | Pass ✅                                                               |||
| 1.4.2a              | not applicable                                                      |||
| 1.4.3a              | Fail 🔴 Contrast of text placeholder is below 4.5:1 | **See #4320** |++|
| 1.4.4a              | Pass ✅                                                               ||| 
| 1.4.5a              | Pass ✅                                                               |||
| 1.4.10a              | Pass ✅                                                               |||
| 1.4.11a              | Fail 🔴 Color Contrast of text inputs' outline not 3:1 to adjecant color AND background of text input not 3:1 to page background |  **See #4321** |++|
| 1.4.12a              | not applicable                                                      |||
| 1.4.13a              | not applicable                                                      |||
| 2.1.1a              | Pass ✅                                                 |||
| 2.1.2a              | Pass ✅                                                 |||
| 2.1.4a              | Pass ✅                                                 |||
| 2.2.1a              | Pass ✅                                                 |||
| 2.2.2a              | Pass ✅                                                 |||
| 2.3.1a              | Pass ✅                                                 |||
| 2.4.1a              | not applicable                                                 |||
| 2.4.2a              | Pass ✅                                                 |||
| 2.4.3a              | Pass ✅                                                 |||
| 2.4.4a              | not applicable                                                      |||
| 2.4.5a              | not applicable                                                      |||
| 2.4.6a              | Fail 🔴 No labels for text inputs                                            |**See #4319**       |++|
| 2.4.7a              | Fail 🔴 No focus style on text inputs.                                                  |**See #4322** |++|
| 2.5.1a              | not applicable                                                      |||
| 2.5.2a              | Pass ✅                                                    |||
| 2.5.3a              | Pass ✅                                                    |||
| 2.5.4a              | not applicable                                                     |||
| 3.1.1a              | Pass ✅ (Login page is in english, but changes to german after login)                                                    |||
| 3.1.2a              | not applicable                                                      |||
| 3.2.1a              | Pass ✅                                                     |||
| 3.2.2a              | Pass ✅                                                     |||
| 3.2.3a              | Pass ✅                                                     |||
| 3.2.4a              | Pass ✅                                                     |||
| 3.3.1a              | Fail 🔴 Error message exists, but is not programmatically connected to invalid text input  |  **See #4323**                                               |+++|
| 3.3.2a              | Fail 🔴 Placeholder attribute is not valid as text input label  | **See #4319**                                               |+++|
| 3.3.3a              | Pass ✅                                                     |||
| 3.3.4a              | not applicable                                                    |||
| 4.1.1a              | Pass ✅                                                     |||
| 4.1.2a              | Pass ✅                                                     |||
| 4.1.3a              | not applicable                                                    |||

**Comment by marcus-herrmann** [comment]
## General Overview when logged in (Scope 1.2)
| BITV test step     | Result                                                                     |
| ------------- |------------------------------------------------|
| 1.1.1a              | partial fail 🟨: Home icon has no accessible name                                                      |
| 1.1.1b              | Pass ✅                                                               |
| 1.1.1c              | not applicable                                                      |
| 1.1.1d              | not applicable                                                      |

**Comment by marcus-herrmann** [comment]
The userflows in Scopes 1.3, 1.4, 1.5 have Scope 1.2 as a starting point and aren't "proper" pages. So please consider the following listing of scope findings "Errors in 1.2 plus 1.3", "...1.4", "...1.5":

## Userflow: Upload a single file (1.3)

**Comment by marcus-herrmann** [comment]
# General findings
- Visible labels of buttons sometimes don't match accessible names, given with `aria-label`
- `aria-label` often/always still English
- `aria-label` applied on `<span>`, they won't work there. Only an element with an role can get an accessible name. `<span>` and `<div>` have implicit `role="none"`, thus, are "role-less"
- When a modal opens, keyboard focus can still reach "behind the modal". The content considered inactive when a modal is open should get `inert` (with polyfill) and `aria-hidden="true"` in the state where the modal is open **OR** a focus trap (ex. https://github.com/posva/focus-trap-vue) should be applied on the modal that the focus can't leave it.

**Comment by marcus-herrmann** [comment]
# Disclaimer / Disclosure
- It is normally not considered ideal that a consultant who worked/consultant on a site/page does the audit (but someone more neutral). Not possible due to timing, and more importantly: this is only the starting audit, not even "Entwicklungsbegleitender Test", nor "Abschließender Test". But I wanted to mention it
- Both WCAG Evaluation Method's / BITV Test Guidance's authors had static pages but not single page apps in mind when writing their documentation. Therefore – and without the opportunity of having a collegue to bounce ideas and decisions off while auditing – I took some educated guesses concerning a few success criteria. There's a good reason that two auditors do the "Abschließender Test" and sync afterwards to find a harmonized verdict
- I'm no certified BITV Tester as of now, can legally speaking offer no reliable advice. But still certified IAAP Web Accessibility Specialist, which is based on WCAG (and BITV 2.0 is also based on WCAG, so the closest I can imagine)
- I have no overview of BITV certified tester and their knowledge regarding the web app niche phoenix is in. Did my best to "translate criteria/test steps" into "web app land", but this is in some regards uncharted land.

**Comment by marcus-herrmann** [comment]
## Userflow: Share a single file (Scope 1.5)
Note: flow starts with activation of button "Add people"


| BITV test step     | Result                                                                     | Issue | Prio Advice |
| ------------- |------------------------------------------------|--|-|
| 1.1.1a              |  not applicable                                           					|||
| 1.1.1b              |  File or folder icon, accordion icon, user avatar has no alternative text ||+++|
| 1.1.1c              |  not applicable                               |||
| 1.1.1d              |  not applicable                                  |||
| 1.2.1a              |   not applicable                             |||
| 1.2.2a              |  not applicable                                |||
| 1.2.3a              |   not applicable                                |||
| 1.2.4a              | not applicable                                 |||
| 1.2.5a              |  not applicable                                |||
| 1.3.1a              |    not applicable                            |||
| 1.3.1b              |   List of "Ausgewählte Personen" in share dialog should be ul                                                            ||++|
| 1.3.1c              |  not applicable                                      |||
| 1.3.1d              |  not applicable                                |||
| 1.3.1e              | Datepicker Ablaufdatum: tabular data, but no table 																			||++|
| 1.3.1f              | not applicable                                                        |||
| 1.3.1g              | not applicable                                    |||
| 1.3.1h              |  not applicable                                                |||
| 1.3.2a              |   not applicable                                            |||
| 1.3.3a              |   not applicable                                    |||
| 1.3.4a              |   not applicable                           |||
| 1.3.5a              |  not applicable                                       |||
| 1.4.1a              |  not applicable                                   |||
| 1.4.2a              |  not applicable                                         |||
| 1.4.3a              | not applicable                                              |||
| 1.4.4a              |  not applicable   | | |
| 1.4.5a              |   not applicable                                        |||
| 1.4.10a              |   not applicable                                           |||
| 1.4.11a              | not applicable  |   ||
| 1.4.12a              |  not applicable                                                                ||
| 1.4.13a              | not applicable  |                                           ||
| 2.1.1a              | Options in "Rolle" dropdown not keyboard accessible. Ablaufdatum Datepicker: not reachable via keyboard       										| |++++|
| 2.1.2a  | not applicable  |                                                            ||
| 2.1.4a              |  not applicable||
| 2.2.1a              | not applicable ||
| 2.2.2a              |  not applicable                                                   |||
| 2.3.1a              |   not applicable                                                           |||
| 2.4.1a              | not applicable| ||
| 2.4.2a              | not applicable |   ||
| 2.4.3a              | Ablaufdatum Datepicker: No focus management                                                         ||++++|
| 2.4.4a              |  not applicable                                                           |||
| 2.4.5a              | not applicable                                                              ||
| 2.4.6a              |  not applicable                                                             ||
| 2.4.7a              | Accordion Focus not visible. Text input focus: Border color change not 3:1 between states (normal/focus)  |  +++|
| 2.5.1a              | not applicable                                                              ||
| 2.5.2a              | not applicable |                                                               |
| 2.5.3a              | not applicable |
| 2.5.4a              | not applicable                                                             ||
| 3.1.1a              | not applicable |                                                    ||
| 3.1.2a              | Close button has English `aria-label` text, although interface language is German |                                                             |+++|
| 3.2.1a              |  not applicable                                                                ||
| 3.2.2a              | not applicable | |
| 3.2.3a              | not applicable                                                                  ||
| 3.2.4a              | not applicable                                                                  ||
| 3.3.1a              | not applicable |       ||
| 3.3.2a              | not applicable                                                                 |||
| 3.3.3a              | not applicable                                                            ||
| 3.3.4a              | not applicable                                                             ||
| 4.1.1a              | not applicable|                                                     ||
| 4.1.2a              | Accordion: Div part has `aria-labelledby`. Icons: wrapping span has `aria-label`. Datepicker in Ablaufdatum uses Divs as buttons  |                                                       |++++|
| 4.1.3a              | not applicable                                        |||

**Comment by marcus-herrmann** [comment]
Finding in 1.5, that I feel I need to emphasize: The Datepicker (context: sharing Links) is really broken. [And as I wrote for ownCloud before, the situation regarding Datepicker accessibility is really grim](https://github.com/owncloud/web/issues/2258#issuecomment-543593897)

**Comment by marcus-herrmann** [comment]
Regarding BITV-Test Update from March, 2021 and test step: 11.7 (https://webtest.bitv-test.de/index.php?a=di&iid=297&s=n). Altough hints for the practice of testing this are still missing in the linked document, I follow the suggestions there and the site performed well. But hesitant to give a final verdict here.

**Comment by marcus-herrmann** [comment]
Not strictly part of Audit #1, but worth a look (and fix), imo: https://github.com/owncloud/web/issues/4867

**Comment by pascalwengerter** [comment]
Closing this since #5018 is merged and we're waiting for feedback (and will open new tickets with the requested changes from the `a11y` audit)
